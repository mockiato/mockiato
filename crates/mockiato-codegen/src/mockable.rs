use crate::diagnostic::DiagnosticBuilder;
use crate::generate::{generate_mock, GenerateMockOptions};
use crate::parse::mockable_attr_parser::{MockableAttrParser, MockableAttrParserImpl};
use crate::parse::trait_decl::TraitDecl;
use crate::result::Error;
use proc_macro2::{Span, TokenStream};
use quote::quote;
use syn::spanned::Spanned;
use syn::{AttributeArgs, Item, ItemTrait};

#[derive(Default)]
pub(crate) struct Mockable;

impl Mockable {
    pub(crate) fn new() -> Self {
        Self::default()
    }

    pub(crate) fn expand(&self, attr: AttributeArgs, item: Item) -> Result<TokenStream, Error> {
        let mockable_attr_parser = MockableAttrParserImpl::new();
        let mockable_attr = mockable_attr_parser.parse(attr)?;
        let item_trait = extract_item_trait(item)?;
        let trait_decl = TraitDecl::parse(item_trait.clone()).map_err(add_note_to_error)?;

        let emit_item_trait = match mockable_attr.remote_trait_path {
            Some(_) => None,
            None => Some(item_trait),
        };

        let generated_mock = generate_mock(
            &trait_decl,
            GenerateMockOptions {
                custom_struct_ident: mockable_attr.name,
                force_static_lifetimes: mockable_attr.force_static_lifetimes,
                custom_trait_path: mockable_attr.remote_trait_path,
            },
        );

        Ok(quote! {
            #emit_item_trait
            #generated_mock
        })
    }
}

fn extract_item_trait(item: Item) -> Result<ItemTrait, Error> {
    match item {
        Item::Trait(item_trait) => Ok(item_trait),
        _ => Err(only_traits_can_be_made_mockable_error(&item)),
    }
}

fn add_note_to_error(error: Error) -> Error {
    error
        .diagnostics
        .into_iter()
        .map(|diagnostic| {
            DiagnosticBuilder::from(diagnostic)
                .note_with_span(Span::call_site(), "Required for mockable traits")
                .build()
        })
        .collect()
}

fn only_traits_can_be_made_mockable_error(item: &Item) -> Error {
    DiagnosticBuilder::error(item.span(), "Only traits can be made mockable")
        .note_with_span(Span::call_site(), "Required because of this attribute")
        .build()
        .into()
}
