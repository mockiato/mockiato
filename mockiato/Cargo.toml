[package]
name = "mockiato"
description = "Minimalistic mocking framework, ready for Rust 2018! ðŸŽ‰"
version = "0.1.1"
authors = [
    "Jan Nils Ferner <jan@myelin.ch>",
    "Ruben Schmidmeister <ruben@myelin.ch>",
    "Jeremy Stucki <jeremy@myelin.ch>",
]
repository = "https://github.com/myelin-ai/mockiato"
homepage = "https://github.com/myelin-ai/mockiato"
documentation = "https://docs.rs/mockiato"
edition = "2018"
license = "MIT"
keywords = [ "test", "mock" ]

[dependencies]
mockiato-codegen = { path = "../mockiato-codegen", version = "0.1.1" }

[dev-dependencies]
compiletest_rs = "0.3.14"
