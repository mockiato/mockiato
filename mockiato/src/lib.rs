mod call;
mod matcher;
pub(crate) mod private;
mod return_value;

pub use self::call::{Call, Calls};
pub use self::matcher::{args, CallMatcher};

// pub use self::return_value::ReturnValue;
