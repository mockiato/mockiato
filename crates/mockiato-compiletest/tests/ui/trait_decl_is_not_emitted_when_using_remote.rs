use mockiato::mockable;
use std::io;

#[mockable(remote = "io::Write")]
trait Write {
    fn write(&mut self, buf: &[u8]) -> io::Result<usize>;

    fn flush(&mut self) -> io::Result<()>;
}

type _AssertTraitDeclIsNotEmitted = dyn Write;

fn main() {
}
