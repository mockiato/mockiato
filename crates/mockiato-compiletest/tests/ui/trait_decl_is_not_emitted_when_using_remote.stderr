error[E0405]: cannot find trait `Write` in this scope
  --> $DIR/trait_decl_is_not_emitted_when_using_remote.rs:11:41
   |
11 | type _AssertTraitDeclIsNotEmitted = dyn Write;
   |                                         ^^^^^ not found in this scope
help: possible candidates are found in other modules, you can import them into scope
   |
1  | use core::fmt::Write;
   |
1  | use std::fmt::Write;
   |
1  | use std::io::Write;
   |
1  | use std::io::prelude::Write;
   |

For more information about this error, try `rustc --explain E0405`.
